algebird-core/src/main/scala/com/twitter/algebird/ExpHist.scala

```diff
@@ -0,0 +1,199 @@
+  def count: Long = bucketSize * length
+
+  // Total number of buckets tracked by the BucketSeq.
+  def length: Int = timestamps.length
+
+  // Add a new timestamp. TODO: NOT sure if it's legit to not require
+  // that the new timestamp be > all existing timestamps.
+  def +(ts: Long): BucketSeq = copy(timestamps = (timestamps :+ ts).sorted)
+
+  def ++(r: BucketSeq): BucketSeq = {
+    require(l.exp == r.exp)
+    copy(timestamps = (l.timestamps ++ r.timestamps).sorted)
+  }
+
+  /**
+   * Remove all timestamps <= the cutoff. Returns:
+   *
+   *  - None if the resulting [[BucketSeq]] is empty,
+   *  - Some(the filtered [[BucketSeq]]) otherwise.
+   */
+  def expire(cutoff: Long): Option[BucketSeq] =
+    Some(copy(timestamps = timestamps.filter(_ > cutoff)))
+      .filter(_.timestamps.isEmpty)
+
+  /**
+   * Returns the number of pairs to drop to get this BucketSeq's final
+   * length <= the supplied limit.
+   */
+  private[this] def pairsToDrop(limit: Int): Int =
+    0 max math.ceil((length - limit) / 2.0).toInt
+
+  /**
+   * if this.length <= limit, returns (this, None).
+   *
+   * else, splits this BucketSeq into two by combining pairs of
+   * timestamps into new buckets until enough timestamps have been
+   * removed that this.length <= limit.
+   *
+   * The new pairs form a new [[BucketSeq]] instance with a doubled
+   * bucketSize.
+   *
+   * The return value in this case is a pair of the new, slimmed-down
+   * current BucketSeq and Some(the new, doubled BucketSeq).
+   */
+  def evolve(limit: Int): (BucketSeq, Option[BucketSeq]) =
+    pairsToDrop(limit) match {
+      case 0 => (this, None)
+      case pairs =>
+        val childTs = (0 until pairs).map(i => timestamps(i * 2))
+        val child = Some(BucketSeq(exp.double, childTs.toVector))
+        copy(timestamps = timestamps.drop(pairs * 2)) -> child
+    }
+
+  /**
+   * Expands this BucketSeq out into a vector of BucketSeq instances
+   * that all have lengths <= the supplied limit. (Used on the oldest
+   * bucket after normalizing a sequence of BucketSeqs.)
+   */
+  def expand(limit: Int): Vector[BucketSeq] =
+    evolve(limit) match {
+      case (bs, None) => Vector(bs)
+      case (bs, Some(remaining)) => remaining.expand(limit) :+ bs
+    }
+
+  // A little janky, but lets us see nice strings of powers of 2.
+  override def toString: String = List.fill(length)(bucketSize).mkString(", ")
+}
+
+object BucketSeq {
+  case class Pow2(exp: Int) extends AnyVal { l =>
+    def double: Pow2 = Pow2(exp + 1)
+    def value: Long = (2 << exp) / 2
+    def *(r: Pow2) = Pow2(l.exp + r.exp)
+  }
+
+  // Returns a singleton bucketseq.
+  def one(timestamp: Long): BucketSeq = BucketSeq(Pow2(0), Vector(timestamp))
+
+  // evolves all BucketSeq instances in `seqs` from newest -> oldest
+  // until every [[BucketSeq]] length is <= the supplied limit.
+  def normalize(limit: Int, seqs: Vector[BucketSeq]): Vector[BucketSeq] = {
```

Comment:
we probably kill this and the evolve functions above once we get the next step of the paper done.